internal/codecutil/encoding.go

```diff
@@ -0,0 +1,118 @@
+// Copyright (C) MongoDB, Inc. 2023-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package codecutil
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"reflect"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/bsoncodec"
+	"go.mongodb.org/mongo-driver/bson/bsonrw"
+	"go.mongodb.org/mongo-driver/x/bsonx/bsoncore"
+)
+
+var ErrNilValue = errors.New("value is nil")
+
+// ErrMapForOrderedArgument is returned when a map with multiple keys is passed
+// to a CRUD method for an ordered parameter
+type ErrMapForOrderedArgument struct {
+	ParamName string
+}
+
+// Error implements the error interface.
+func (e ErrMapForOrderedArgument) Error() string {
+	return fmt.Sprintf("multi-key map passed in for ordered parameter %v",
+		e.ParamName)
+}
+
+// MarshalError is returned when attempting to transform a value into a document
+// results in an error.
+type MarshalError struct {
+	Value interface{}
+	Err   error
+}
+
+// Error implements the error interface.
+func (e MarshalError) Error() string {
+	return fmt.Sprintf("cannot transform type %s to a BSON Document: %v",
+		reflect.TypeOf(e.Value), e.Err)
+}
+
+// EncoderFn is used to functionally construct an encoder for marshaling values.
+type EncoderFn func(io.Writer, *bsoncodec.Registry) (*bson.Encoder, error)
```

Comment:
Optional: It's kinda confusing that part of the configuration of the `Encoder` is handled when creating the `EncoderFn` (applying the `BSONOptions`) and part is handled when calling it (applying the `Registry`). It could be simplified by only allowing callers to pass in the destination `io.Writer`, requiring the behavior of the `EncoderFn` to be defined at creation time.

E.g. updated interface:
```go
type EncoderFn func(io.Writer) (*bson.Encoder, error)
```

Reply:
Oh, great idea!